plugin-maven/src/main/java/com/diffplug/spotless/maven/generic/LicenseHeader.java

```diff
@@ -41,24 +44,50 @@ public final FormatterStep newFormatterStep(FormatterStepConfig config) {
 		if (delimiterString == null) {
 			throw new IllegalArgumentException("You need to specify 'delimiter'.");
 		}
-
 		if (file != null ^ content != null) {
-			FormatterStep step = file != null
-					? createStepFromFile(config, delimiterString)
-					: createStepFromContent(delimiterString);
+			FormatterStep unfiltered;
+			if ("true".equals(config.spotlessSetLicenseHeaderYearsFromGitHistory().orElse(""))) {
+				unfiltered = FormatterStep.createNeverUpToDateLazy(LicenseHeaderStep.name(), () -> {
+					boolean updateYear = false; // doesn't matter
+					LicenseHeaderStep step = new LicenseHeaderStep(readFileOrContent(config), delimiterString, LicenseHeaderStep.defaultYearDelimiter(), updateYear);
+					return new FormatterFunc() {
+						@Override
+						public String apply(String input, File source) throws Exception {
+							return step.setLicenseHeaderYearsFromGitHistory(input, source);
+						}
 
-			return step.filterByFile(LicenseHeaderStep.unsupportedJvmFilesFilter());
+						@Override
+						public String apply(String input) throws Exception {
+							throw new UnsupportedOperationException();
+						}
+					};
+				});
+			} else {
+				unfiltered = FormatterStep.createLazy(LicenseHeaderStep.name(), () -> {
+					// by default, we should update the year if the user is using ratchetFrom
+					boolean updateYear = config.getRatchetFrom().isPresent();
+					String header;
+					if (content != null) {
```

Comment:
Maybe use `readFileOrContent` instead of this if-else?

Reply:
Another good catch.